### PR TITLE
Module6

### PR DESCRIPTION
**Pull request by KAVTARADZELUKA**

### CODE REVIEW

File: src/main/java/com/gym/gymsystem/config/GlobalCorsConfig.java [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+import org.springframework.web.servlet.config.annotation.CorsRegistry;
+import org.springframework.web.servlet.config.annotation.WebMvcConfigurer;
+
+@Configuration
+public class GlobalCorsConfig implements WebMvcConfigurer {
+    @Override
+    public void addCorsMappings(CorsRegistry registry) {
+        registry.addMapping("/**")
+                .allowedOrigins("*")
+                .allowedMethods("GET", "POST", "PUT", "DELETE", "OPTIONS")
```
Comment (HannaVatlina) [inline src/main/java/com/gym/gymsystem/config/GlobalCorsConfig.java]: What about "PATCH" method?

File: src/main/java/com/gym/gymsystem/controller/TrainerController.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+            return ResponseEntity.status(HttpStatus.CREATED).body(response);
+        });
+    }
+
+    @GetMapping("/{findUsername}")
+    public ResponseEntity<TrainerProfileResponse> getTrainerProfile(
+            @PathVariable("findUsername") String findUsername) {
+        if (!authorizationService.isAdmin()  && !authorizationService.isAuthenticatedUser(findUsername)) {
+            throw new CustomAccessDeniedException("You do not have permission to get this trainer's profile.");
+        }
```
Comment (HannaVatlina) [inline src/main/java/com/gym/gymsystem/controller/TrainerController.java]: Is this required if you have this configuration `.requestMatchers(HttpMethod.GET,"/api/trainer/*").hasAnyRole("ADMIN", "TRAINER")`?

File: src/main/java/com/gym/gymsystem/config/SecurityConfig.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+                        .requestMatchers(HttpMethod.GET,"/api/trainer/*").hasAnyRole("ADMIN", "TRAINER")
+                        .requestMatchers(HttpMethod.PUT,"/api/trainer/*").hasAnyRole("ADMIN", "TRAINER")
+                        .requestMatchers(HttpMethod.PATCH,"/api/trainer/status/*").hasAnyRole("ADMIN")
+
+                        .requestMatchers(HttpMethod.PUT,"/api/training/trainers").hasAnyRole("ADMIN", "TRAINEE")
+                        .requestMatchers(HttpMethod.GET,"/api/training/trainer").hasAnyRole("ADMIN", "TRAINER")
+                        .requestMatchers(HttpMethod.POST,"/api/training").hasAnyRole("ADMIN", "TRAINER")
+                        .requestMatchers("/actuator").permitAll()
+                        .requestMatchers("/actuator/**").permitAll()
+                        .anyRequest().authenticated()
```
Comment (HannaVatlina) [inline src/main/java/com/gym/gymsystem/config/SecurityConfig.java:69]: Do I need authorization to access swagger UI?

File: src/main/java/com/gym/gymsystem/controller/TrainingTypeController.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+@RequestMapping("/api/training-types")
+public class TrainingTypeController {
+    private final TrainingTypeService trainingTypeService;
+
+    public TrainingTypeController(TrainingTypeService trainingTypeService) {
+        this.trainingTypeService = trainingTypeService;
+    }
+
+    @GetMapping
+    public ResponseEntity<List<TrainingType>> getAllTrainingTypes() {
```
Comment (HannaVatlina) [inline src/main/java/com/gym/gymsystem/controller/TrainingTypeController.java:22]: This endpoint is under authorization, was this done on purpose?

File: src/main/java/com/gym/gymsystem/config/SecurityConfig.java [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    @Bean
+    public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {
+        http
+                .csrf(AbstractHttpConfigurer::disable)
+                .authorizeHttpRequests(auth -> auth
+                        .requestMatchers(HttpMethod.POST, "/api/trainee").permitAll()
+                        .requestMatchers(HttpMethod.POST, "/api/trainer").permitAll()
+                        .requestMatchers( "/api/auth/change-password").permitAll()
+
+                        .requestMatchers(HttpMethod.PUT,"/api/trainee/*").hasAnyRole("ADMIN", "TRAINEE")
```
Comment (HannaVatlina) [inline src/main/java/com/gym/gymsystem/config/SecurityConfig.java]: a lot of duplicate strings ("ADMIN", "TRAINEE"), you can create enum for this

File: src/main/java/com/gym/gymsystem/config/SecurityConfig.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    @Bean
+    public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {
+        http
+                .csrf(AbstractHttpConfigurer::disable)
+                .authorizeHttpRequests(auth -> auth
+                        .requestMatchers(HttpMethod.POST, "/api/trainee").permitAll()
+                        .requestMatchers(HttpMethod.POST, "/api/trainer").permitAll()
+                        .requestMatchers( "/api/auth/change-password").permitAll()
```
Comment (HannaVatlina) [inline src/main/java/com/gym/gymsystem/config/SecurityConfig.java]: `/api/ `is repeatable in your endpoints, you can use '`server.servlet.context-path`' for this

File: src/main/java/com/gym/gymsystem/controller/TrainingController.java [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+        if (!authorizationService.isAdmin() && !authorizationService.isTrainer() && !authorizationService.isAuthenticatedUser(findUsername)) {
+            throw new CustomAccessDeniedException("You do not have permission to get this trainee's not assigned trainers.");
+        }
+        List<TrainerInfo> trainerInfos = trainingService.getTrainersNotAssignedToTrainee(findUsername);
+        return ResponseEntity.ok(trainerInfos);
+    }
+
+    @PutMapping("/trainers")
+    public ResponseEntity<List<TrainerInfo>> updateTraineeTrainers(
+            @RequestParam("findUsername") String findUsername,
```
Comment (HannaVatlina) [inline src/main/java/com/gym/gymsystem/controller/TrainingController.java]: I guess this is trainer username. if so, it better to use path param for this and named like trainerUsername

File: src/main/java/com/gym/gymsystem/controller/TrainingController.java [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+                ))
+                .distinct()
+                .toList();
+
+        return ResponseEntity.ok(trainerInfos);
+    }
+
+    @GetMapping("/trainee")
+    public ResponseEntity<List<TrainingInfo>> getTraineeTrainings(
+            @RequestParam("traineeUsername") String traineeUsername,
```
Comment (HannaVatlina) [inline src/main/java/com/gym/gymsystem/controller/TrainingController.java]: same here if this is a required field and you want to return all trainings for a trainee then use path parameter

File: src/main/java/com/gym/gymsystem/service/LoginAttemptService.java [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    private static final long LOCK_TIME_DURATION = 5; // 5 minutes
+
+    private final LoginAttemptRepository loginAttemptRepository;
+
+    @Autowired
+    public LoginAttemptService(LoginAttemptRepository loginAttemptRepository) {
+        this.loginAttemptRepository = loginAttemptRepository;
+    }
+
+    public void loginFailed(String username) {
```
Comment (HannaVatlina) [inline src/main/java/com/gym/gymsystem/service/LoginAttemptService.java:25]: Should it be executed in a transaction?

File: src/main/java/com/gym/gymsystem/service/LoginAttemptService.java [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+        loginAttempt.setLastAttempt(LocalDateTime.now());
+
+        if (loginAttempt.getAttempts() >= MAX_ATTEMPTS) {
+            loginAttempt.setLockedUntil(LocalDateTime.now().plusMinutes(LOCK_TIME_DURATION));
+        }
+
+        loginAttemptRepository.save(loginAttempt);
+    }
+
+    public void loginSucceeded(String username) {
```
Comment (HannaVatlina) [inline src/main/java/com/gym/gymsystem/service/LoginAttemptService.java:41]: Should it be executed in a transaction?

File: src/main/java/com/gym/gymsystem/service/TrainingService.java [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+    }
+
+    public List<Training> findTrainingsByTrainerAndCriteria(String findUsername, LocalDateTime fromDate, LocalDateTime toDate, String trainerName, String trainingType) {
+        return trainingRepository.findTrainingsByTrainerAndCriteria(findUsername, fromDate, toDate, trainerName, trainingType);
+    }
+
+    public List<TrainerInfo> getTrainersNotAssignedToTrainee(String findUsername) {
+        List<Trainer> availableTrainers = trainingRepository.findTrainersNotAssignedToTrainee(findUsername);
+
+        return availableTrainers.stream()
```
Comment (HannaVatlina) [inline src/main/java/com/gym/gymsystem/service/TrainingService.java:66]: Should we return Trainers who are disable?

File: src/main/java/com/gym/gymsystem/util/JwtTokenUtil.java [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+
+import io.jsonwebtoken.Claims;
+import io.jsonwebtoken.Jwts;
+import io.jsonwebtoken.SignatureAlgorithm;
+import org.springframework.stereotype.Component;
+
+import javax.crypto.KeyGenerator;
+import javax.crypto.SecretKey;
+import java.security.NoSuchAlgorithmException;
+import java.util.Date;
```
Comment (HannaVatlina) [inline src/main/java/com/gym/gymsystem/util/JwtTokenUtil.java:13]: work with date from java.time package

File: src/main/java/com/gym/gymsystem/controller/TrainingTypeController.java (hunk 4 above)
Comment (KAVTARADZELUKA) [reply src/main/java/com/gym/gymsystem/controller/TrainingTypeController.java:22]: I think it is not necessary to authorize to see this information, the guest should also be able to see what kind of trainings we have

Comment (HannaVatlina) [reply src/main/java/com/gym/gymsystem/controller/TrainingTypeController.java:22]: you are right, but currently authorization is needed.
You have this configuration .anyRequest().authenticated()

Comment (KAVTARADZELUKA) [reply src/main/java/com/gym/gymsystem/controller/TrainingTypeController.java:22]: You're right, I missed it, I'll fix it, thanks

**Pull request merged**